src/hotspot/share/oops/methodData.cpp

```diff
@@ -1821,3 +1821,13 @@ void MethodData::clean_weak_method_links() {
   clean_extra_data(&cl);
   verify_extra_data_clean(&cl);
 }
+
+void MethodData::deallocate_contents(ClassLoaderData* loader_data) {
+  release_C_heap_structures();
+}
+
+void MethodData::release_C_heap_structures() {
+#if INCLUDE_JVMCI
+  FailedSpeculation::free_failed_speculations(get_failed_speculations_address());
```

Comment:
I don't think this can be called twice either. Looking at FailedSpeculation::free_failed_speculations it doesn't appear to have logic to detect freeing twice.

Reply:
Yes.